cmd/zrepl/zrepl.c

```diff
@@ -734,6 +1075,137 @@ uzfs_zvol_mgmt_thread(void *arg)
 	zk_thread_exit();
 }
 
+
+static int
+uzfs_zvol_rebuild_scanner_callback(off_t offset, size_t len,
+    blk_metadata_t *metadata, zvol_state_t *zv, void *args)
+{
+	zvol_io_hdr_t	hdr;
+	zvol_io_cmd_t	*zio_cmd;
+	zvol_rebuild_t  *warg;
+	zvol_info_t	*zinfo;
+
+	warg = (zvol_rebuild_t *)args;
+	zinfo = warg->zinfo;
+
+	hdr.version = REPLICA_VERSION;
+	hdr.opcode = ZVOL_OPCODE_READ;
+	hdr.io_seq = metadata->io_num;
+	hdr.offset = offset;
+	hdr.len = len;
+	hdr.flags = ZVOL_OP_FLAG_REBUILD;
+	hdr.status = ZVOL_OP_STATUS_OK;
+	printf("IO number for rebuild %ld\n", metadata->io_num);
+	zio_cmd = zio_cmd_alloc(&hdr, warg->fd);
+	/* Take refcount for uzfs_zvol_worker to work on it */
+	uzfs_zinfo_take_refcnt(zinfo, B_FALSE);
+	zio_cmd->zv = zinfo;
+	uzfs_zvol_worker(zio_cmd);
+	return (0);
+}
+
+/*
+ * Rebuild scanner function which after receiving
+ * vol_name and IO number, will scan metadata and
+ * read data and send across.
+ */
+static void
+uzfs_zvol_rebuild_scanner(void *arg)
+{
+	int		fd;
+	zvol_info_t	*zinfo = NULL;
+	zvol_io_hdr_t	hdr;
+	int 		rc = 0;
+	zvol_rebuild_t  warg;
+	char 		*name;
+	blk_metadata_t	metadata;
+	uint64_t	rebuild_req_offset;
+	uint64_t	rebuild_req_len;
+	zvol_io_cmd_t	*zio_cmd;
+
+	fd = *(int *)arg;
+	free(arg);
```

Comment:
better if we free mem by kmem_free if it was allocated by kmem_alloc. Note that the allocation and deallocation can be avoided entirely if we store int directly into void pointer.